### PR TITLE
Simple Attachment Storage implementation using Requery/H2 database

### PR DESCRIPTION
**Pull request by polybioz**
Recreated a pull request #246 after strange issues with rebasing.

### CODE REVIEW

**Comment by mikehearn** [comment]
I'll review this after Jose is happy.

**Comment by josecoll** [comment]
Schema generation in Node module:
I experienced a lot of strange behaviour with Kapt executing in the Node module, but perhaps this has gone away with later versions of the Kapt processor plugin.

File: node/src/main/kotlin/net/corda/node/services/persistence/NodeRqAttachmentService.kt [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+            try {
+                session.select(AttachmentEntity::class)
+                        .where(AttachmentEntity.ATT_ID.eq(id))
+                        .get()
+                        .single()
+            } catch (e: NoSuchElementException) {
+                null
+            }
+        } ?: return null
+
```
Comment (polybioz) [inline node/src/main/kotlin/net/corda/node/services/persistence/NodeRqAttachmentService.kt]: The original service `NodeAttachmentService` returns null and it is expected in several places:
* `FetchDataHandler::call`
* `ResolveTransactionsFlow::fetchMissingAttachments`
* `WireTransactionSerializer::read`
* `WireTransaction::toLedgerTransaction`
* `CordaRPCOpsImpl::attachmentExists`

**Comment by mikehearn** [comment]
@josecoll is the ball in your court here?

File: node/build.gradle [hunk 2]
```diff
@@ -1,4 +1,5 @@
 apply plugin: 'kotlin'
+apply plugin: 'kotlin-kapt'
 apply plugin: 'java'
```
Comment (josecoll) [inline node/build.gradle]: Please remove

File: node/build.gradle [hunk 3]
```diff
@@ -41,6 +42,9 @@ sourceSets {
         }
     }
     main {
+        kotlin {
+            srcDir "${buildDir}/generated/source/kapt/main/"
+        }
```
Comment (josecoll) [inline node/build.gradle]: Please remove 45 - 47

File: core/src/test/kotlin/net/corda/core/flows/ResolveTransactionsFlowTest.kt [hunk 4]
```diff
@@ -134,13 +134,17 @@ class ResolveTransactionsFlowTest {
 
     @Test
     fun attachment() {
-        val id = a.services.storageService.attachments.importAttachment("Some test file".toByteArray().opaque().open())
+        val id = databaseTransaction(a.database) {
+            a.services.storageService.attachments.importAttachment("Some test file".toByteArray().opaque().open())
+        }
```
Comment (josecoll) [inline core/src/test/kotlin/net/corda/core/flows/ResolveTransactionsFlowTest.kt]: Please add a TODO here to revisit transaction handling (eg. this operation should not require an explicit transaction).
I will raise a separate story for investigating the inconsistent behaviour you encountered.

File: core/src/test/kotlin/net/corda/core/flows/ResolveTransactionsFlowTest.kt [hunk 5]
```diff
@@ -134,13 +134,17 @@ class ResolveTransactionsFlowTest {
...
         val stx2 = makeTransactions(withAttachment = id).second
         val p = ResolveTransactionsFlow(stx2, a.info.legalIdentity)
         val future = b.services.startFlow(p).resultFuture
         net.runNetwork()
         future.getOrThrow()
-        assertNotNull(b.services.storageService.attachments.openAttachment(id))
+        databaseTransaction(b.database) {
+            assertNotNull(b.services.storageService.attachments.openAttachment(id))
+        }
     }
```
Comment (josecoll) [inline core/src/test/kotlin/net/corda/core/flows/ResolveTransactionsFlowTest.kt]: Please add a TODO here to revisit transaction handling (eg. this operation should not require an explicit transaction).
I will raise a separate story for investigating the inconsistent behaviour you encountered.

File: node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt [hunk 6]
```diff
@@ -101,9 +101,24 @@ class CordaRPCOpsImpl(
...
     }
 
-    override fun attachmentExists(id: SecureHash) = services.storageService.attachments.openAttachment(id) != null
-    override fun openAttachment(id: SecureHash) = services.storageService.attachments.openAttachment(id)!!.open()
-    override fun uploadAttachment(jar: InputStream) = services.storageService.attachments.importAttachment(jar)
+    override fun attachmentExists(id: SecureHash): Boolean {
+        return databaseTransaction(database){
+            services.storageService.attachments.openAttachment(id) != null
+        }
+    }
```
Comment (josecoll) [inline node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt]: Please add a TODO here to revisit transaction handling (eg. this operation should not require an explicit transaction).
I will raise a separate story for investigating the inconsistent behaviour you encountered.

File: node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt [hunk 7]
```diff
@@ -101,9 +101,24 @@ class CordaRPCOpsImpl(
...
+        return databaseTransaction(database) {
+            services.storageService.attachments.openAttachment(id)!!.open()
+        }
+    }
+
+    override fun uploadAttachment(jar: InputStream): SecureHash {
+        return databaseTransaction(database){
+            services.storageService.attachments.importAttachment(jar)
+        }
+    }
```
Comment (josecoll) [inline node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt]: Please add a TODO here to revisit transaction handling (eg. this operation should not require an explicit transaction).
I will raise a separate story for investigating the inconsistent behaviour you encountered.

File: node/src/test/kotlin/net/corda/node/messaging/AttachmentTests.kt [hunk 8]
```diff
@@ -93,19 +109,32 @@ class AttachmentTests {
...
                 }
             }
         }, true, null, null, ServiceInfo(NetworkMapService.type), ServiceInfo(SimpleNotaryService.type))
         val n1 = network.createNode(n0.info.address)
 
+        val attachment = fakeAttachment()
         // Insert an attachment into node zero's store directly.
-        val id = n0.storage.attachments.importAttachment(ByteArrayInputStream(fakeAttachment()))
+        val id = databaseTransaction(n0.database)
+        {
```
Comment (josecoll) [inline node/src/test/kotlin/net/corda/node/messaging/AttachmentTests.kt]: move { to line above

**Comment by josecoll** [comment]
Looks good to me.
Over to you @mikehearn

File: core/src/main/kotlin/net/corda/core/schemas/requery/converters/BlobConverter.kt [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+    override fun getPersistedType(): Class<Blob> { return Blob::class.java }
+
+    override fun getPersistedSize(): Int? { return null }
+
+    override fun convertToPersisted(value: ByteArray?): Blob? {
+        return value?.let { SerialBlob(value) }
+    }
+
+    override fun convertToMapped(type: Class<out ByteArray>?, value: Blob?): ByteArray? {
```
Comment (polybioz) [inline core/src/main/kotlin/net/corda/core/schemas/requery/converters/BlobConverter.kt]: Nulls can passed and  returned. Built-in Requery converters have null checks too.

File: node/src/main/kotlin/net/corda/node/services/persistence/NodeAttachmentService.kt [hunk 10]
```diff
@@ -19,73 +26,68 @@ import java.util.jar.JarInputStream
...
-    // Just count all non-directories in the attachment store, and assume the admin hasn't dumped any junk there.
-    private fun countAttachments() = storePath.list { it.filter { it.isRegularFile() }.count() }
-
-    /**
-     * If true, newly inserted attachments will be unzipped to a subdirectory of the [storePath]. This is intended for
-     * human browsing convenience: the attachment itself will still be the file (that is, edits to the extracted directory
-     * will not have any effect).
-     */
-    @Volatile var automaticallyExtractAttachments = false
+    override var automaticallyExtractAttachments = false
```
Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/services/persistence/NodeAttachmentService.kt]: I think this still needs to be marked as volatile

**Comment by mikehearn** [comment]
@josecoll Can you take this PR over and finish it off? It needs a rebase and I think the volatile annotation needs to be restored. Other than that it looked OK to me.

**Comment by josecoll** [comment]
Sure - will complete the final PR request & rebase tomorrow.

**Comment by polybioz** [comment]
I restored volatile annotation and rebased the PR.

**Comment by mikehearn** [comment]
Ah, thanks! Let's squash and merge.

**Comment by mikehearn** [comment]
@polybioz can you squash the commits into one? (using git rebase again) - there's no need for our git history to contain the history of the code review too.

**Comment by polybioz** [comment]
@mikehearn done

**Comment by mikehearn** [comment]
Thanks for the great contribution!

**Pull request merged**